File: test/language/sql_constraints/java/sql_constraints/field_constraints/FieldConstraintsTest.java
package sql_constraints.field_constraints;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.CoreMatchers.containsString;
import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.IOException;
import java.sql.SQLException;
import java.util.Arrays;

import test_utils.FileUtil;
import test_utils.JdbcUtil;

import sql_constraints.TestDb;

public class FieldConstraintsTest
{
    @BeforeAll
    public static void init()
    {
        JdbcUtil.registerJdbc();
    }

    @BeforeEach
    public void setUp() throws IOException, SQLException
    {
        FileUtil.deleteFileIfExists(file);
        database = new TestDb(file.toString());
        database.createSchema();
    }

    @AfterEach
    public void tearDown() throws SQLException
    {
        if (database != null)
        {
            database.close();
            database = null;
        }
    }

    @Test
    public void withoutSql() throws IOException, SQLException
    {
        final FieldConstraintsTable fieldConstraintsTable = database.getFieldConstraintsTable();
        final FieldConstraintsTableRow row = new FieldConstraintsTableRow();
        fillRow(row);
        row.setNullWithoutSql();
        fieldConstraintsTable.write(Arrays.asList(row));
    }

    @Test
    public void sqlNotNull() throws IOException, SQLException
    {
        final FieldConstraintsTable fieldConstraintsTable = database.getFieldConstraintsTable();
        final FieldConstraintsTableRow row = new FieldConstraintsTableRow();
        fillRow(row);
        row.setNullSqlNotNull();
        final SQLException thrown =
                assertThrows(SQLException.class, () -> fieldConstraintsTable.write(Arrays.asList(row)));

        assertThat(thrown.getMessage(),
                containsString("NOT NULL constraint failed: fieldConstraintsTable.sqlNotNull"));
    }

    @Test
    public void sqlDefaultNull() throws IOException, SQLException
    {
        final FieldConstraintsTable fieldConstraintsTable = database.getFieldConstraintsTable();
        final FieldConstraintsTableRow row = new FieldConstraintsTableRow();
        fillRow(row);
        row.setNullSqlDefaultNull();
        fieldConstraintsTable.write(Arrays.asList(row));
    }

    @Test
    public void sqlCheckConstant() throws IOException, SQLException
    {
        final FieldConstraintsTable fieldConstraintsTable = database.getFieldConstraintsTable();
        final FieldConstraintsTableRow row = new FieldConstraintsTableRow();
        fillRow(row);
        row.setSqlCheckConstant(WRONG_CONSTRAINTS_CONSTANT);
        final SQLException thrown =
                assertThrows(SQLException.class, () -> fieldConstraintsTable.write(Arrays.asList(row)));

        assertThat(thrown.getMessage(), containsString("CHECK constraint failed: fieldConstraintsTable"));
    }

    @Test
    public void sqlCheckImportedConstant() throws IOException, SQLException
    {

        final FieldConstraintsTable fieldConstraintsTable = database.getFieldConstraintsTable();
        final FieldConstraintsTableRow row = new FieldConstraintsTableRow();
        fillRow(row);
        row.setSqlCheckImportedConstant(WRONG_IMPORTED_CONSTRAINTS_CONSTANT);
        final SQLException thrown =
                assertThrows(SQLException.class, () -> fieldConstraintsTable.write(Arrays.asList(row)));

        assertThat(thrown.getMessage(), containsString("CHECK constraint failed: fieldConstraintsTable"));
    }

    @Test
    public void sqlCheckUnicodeEscape() throws IOException, SQLException
    {
        final FieldConstraintsTable fieldConstraintsTable = database.getFieldConstraintsTable();
        final FieldConstraintsTableRow row = new FieldConstraintsTableRow();
        fillRow(row);
        row.setSqlCheckUnicodeEscape(WRONG_UNICODE_ESCAPE_CONST);
        final SQLException thrown =
                assertThrows(SQLException.class, () -> fieldConstraintsTable.write(Arrays.asList(row)));

        assertThat(thrown.getMessage(), containsString("CHECK constraint failed: fieldConstraintsTable"));
    }

    @Test
    public void sqlCheckHexEscape() throws IOException, SQLException
    {
        final FieldConstraintsTable fieldConstraintsTable = database.getFieldConstraintsTable();
        final FieldConstraintsTableRow row = new FieldConstraintsTableRow();
        fillRow(row);
        row.setSqlCheckHexEscape(WRONG_HEX_ESCAPE_CONST);
        final SQLException thrown =
                assertThrows(SQLException.class, () -> fieldConstraintsTable.write(Arrays.asList(row)));

        assertThat(thrown.getMessage(), containsString("CHECK constraint failed: fieldConstraintsTable"));
    }

    @Test
    public void sqlCheckOctalEscape() throws IOException, SQLException
    {
        final FieldConstraintsTable fieldConstraintsTable = database.getFieldConstraintsTable();
        final FieldConstraintsTableRow row = new FieldConstraintsTableRow();
        fillRow(row);
        row.setSqlCheckOctalEscape(WRONG_OCTAL_ESCAPE_CONST);
        final SQLException thrown =
                assertThrows(SQLException.class, () -> fieldConstraintsTable.write(Arrays.asList(row)));

        assertThat(thrown.getMessage(), containsString("CHECK constraint failed: fieldConstraintsTable"));
    }

    private void fillRow(FieldConstraintsTableRow row)
    {
        row.setWithoutSql(1);
        row.setSqlNotNull(1);
        row.setSqlDefaultNull(1);
        row.setSqlCheckConstant(1);
        row.setSqlCheckImportedConstant(1);
        row.setSqlCheckUnicodeEscape(UNICODE_ESCAPE_CONST);
        row.setSqlCheckHexEscape(HEX_ESCAPE_CONST);
        row.setSqlCheckOctalEscape(OCTAL_ESCAPE_CONST);
    }

    private static final String FILE_NAME = "field_constraints_test.sqlite";

    private static final short UNICODE_ESCAPE_CONST = 1;
    private static final short HEX_ESCAPE_CONST = 2;
    private static final short OCTAL_ESCAPE_CONST = 3;

    private static final int WRONG_CONSTRAINTS_CONSTANT = 124;
    private static final long WRONG_IMPORTED_CONSTRAINTS_CONSTANT = 322;

    private static final short WRONG_UNICODE_ESCAPE_CONST = 0;
    private static final short WRONG_HEX_ESCAPE_CONST = 0;
    private static final short WRONG_OCTAL_ESCAPE_CONST = 0;

    private final File file = new File(FILE_NAME);
    private TestDb database = null;
}

File: test/extensions/language/sql_constraints/python/FieldConstraintsTest.py
import os
import apsw

import SqlConstraints


class FieldConstraintsTest(SqlConstraints.TestCaseWithDb):
    def setUp(self):
        if os.path.exists(self.dbFileName):
            os.remove(self.dbFileName)
        self._database = self.api.TestDb.from_file(self.dbFileName)
        self._database.create_schema()
        self._constraintsTable = self._database.field_constraints_table

    def tearDown(self):
        self._database.close()

    def testWithoutSql(self):
        rowDict = self._createRowDict()
        rowDict["withoutSql"] = None
        self._writeRowDict(rowDict)

    def testSqlNotNull(self):
        rowDict = self._createRowDict()
        rowDict["sqlNotNull"] = None
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
        self.assertTrue(
            "NOT NULL constraint failed: fieldConstraintsTable.sqlNotNull" in str(context.exception),
            str(context.exception),
        )

    def testSqlDefaultNull(self):
        rowDict = self._createRowDict()
        rowDict["sqlDefaultNull"] = None
        self._writeRowDict(rowDict)

    def testSqlCheckConstant(self):
        rowDict = self._createRowDict()
        rowDict["sqlCheckConstant"] = self.WRONG_CONSTRAINTS_CONSTANT
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
        self.assertTrue(
            "CHECK constraint failed: sqlCheckConstant" in str(context.exception), str(context.exception)
        )

    def testSqlCheckImportedConstant(self):
        rowDict = self._createRowDict()
        rowDict["sqlCheckImportedConstant"] = self.WRONG_IMPORTED_CONSTRAINTS_CONSTANT
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
        self.assertTrue(
            "CHECK constraint failed: sqlCheckImportedConstant" in str(context.exception),
            str(context.exception),
        )

    def testSqlCheckUnicodeEscape(self):
        rowDict = self._createRowDict()
        rowDict["sqlCheckUnicodeEscape"] = self.WRONG_UNICODE_ESCAPE_CONST
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
        self.assertTrue(
            "CHECK constraint failed: sqlCheckUnicodeEscape" in str(context.exception), str(context.exception)
        )

    def testSqlCheckHexEscape(self):
        rowDict = self._createRowDict()
        rowDict["sqlCheckHexEscape"] = self.WRONG_HEX_ESCAPE_CONST
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
        self.assertTrue(
            "CHECK constraint failed: sqlCheckHexEscape" in str(context.exception), str(context.exception)
        )

    def testSqlCheckOctalEscape(self):
        rowDict = self._createRowDict()
        rowDict["sqlCheckOctalEscape"] = self.WRONG_OCTAL_ESCAPE_CONST
        with self.assertRaises(apsw.ConstraintError) as context:
            self._writeRowDict(rowDict)
        self.assertTrue(
            "CHECK constraint failed: sqlCheckOctalEscape" in str(context.exception), str(context.exception)
        )

    def _createRowDict(self):
        return {
            "primaryKey": 1,
            "withoutSql": 1,
            "sqlNotNull": 1,
            "sqlDefaultNull": 1,
            "sqlCheckConstant": 1,
            "sqlCheckImportedConstant": 1,
            "sqlCheckUnicodeEscape": self.UNICODE_ESCAPE_CONST,
            "sqlCheckHexEscape": self.HEX_ESCAPE_CONST,
            "sqlCheckOctalEscape": self.OCTAL_ESCAPE_CONST,
        }

    def _writeRowDict(self, rowDict):
        row = [
            (
                rowDict["primaryKey"],
                rowDict["withoutSql"],
                rowDict["sqlNotNull"],
                rowDict["sqlDefaultNull"],
                rowDict["sqlCheckConstant"],
                rowDict["sqlCheckImportedConstant"],
                rowDict["sqlCheckUnicodeEscape"],
                rowDict["sqlCheckHexEscape"],
                rowDict["sqlCheckOctalEscape"],
            )
        ]
        self._constraintsTable.write(row)

    UNICODE_ESCAPE_CONST = 1
    HEX_ESCAPE_CONST = 2
    OCTAL_ESCAPE_CONST = 3

    WRONG_CONSTRAINTS_CONSTANT = 124
    WRONG_IMPORTED_CONSTRAINTS_CONSTANT = 322

    WRONG_UNICODE_ESCAPE_CONST = 0
    WRONG_HEX_ESCAPE_CONST = 0
    WRONG_OCTAL_ESCAPE_CONST = 0
